### PR TITLE
Overhaul of bluetooth connection to accommodate for user-friendliness of both windows and mac installations

### PR DESCRIPTION
**Pull request by rdvhoorn**
Previously, winreg was used to initiate a Bluetooth connection from windows devices. This is no longer necessary. 

-Winreg completely removed from code
-Added automatic Bluetooth pairing capability for MAC (not tested on Linux)
-Moved `distribution` parameter to definitions.py
-Added `RUNNING_ON_MAC` to definitions.py
-Rewrote README
-Reworked initialization of server and routes.
-Reworked `requirements.txt` to only show needed packages.

Note: the changes in the Routes.py file seem very drastic but is only like indented once as it is now encompassed in an outer 'initialization' function which is called by the app.py script.

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -1,41 +1,44 @@
...
-
-## Running for MAC
-You can find some general information about running the software on OSX in the [general documentation on the Drive](https://docs.google.com/document/d/1RFTCGX0A7BAd-lvrdRqAtj_ZibQ0j-yJanyfVRDa4ME/edit).
-
+4. Create a folder for the virtual environment called venv (`<root>\venv\`)
+5. Open a terminal (admin mode) in the root folder.
+6. Install virtualenv: `pip install virtualenv`
+7. Go to the venv directory: `cd venv`
+8. Create a new venv: `py -m venv backend`
+9. Activate the environment: `.\backend\Scripts\activate`
```
Comment (hendriklohse) [inline README.md:14]: For mac this is: backend/bin/activate

File: README.md [hunk 2]
```diff
@@ -1,41 +1,44 @@
...
+10. Check if activation worked: `where python` should return `...\<root>\venv\backend\Scripts\python.exe`
```
Comment (hendriklohse) [inline README.md:15]: For mac 'where' is not a command. I think the more general approach here is to do:
echo $VIRTUAL_ENV
This will return '.../<root>/venvs/backend'
Once you run deactivate / leave the virtual environment, the $VIRTUAL_ENV variable will be cleared/empty.

source: https://stackoverflow.com/questions/1871549/determine-if-python-is-running-inside-virtualenv
The answer from @florisla (4th) gives a nice overview of all possible methods.

File: README.md [hunk 3]
```diff
@@ -1,41 +1,44 @@
 # backend
```
Comment (hendriklohse) [inline README.md:1]: General remark:
There is no consistency regarding the use of '\' (windows) and '/' (mac). If we want one general installation I'd opt for '\'.

File: README.md [hunk 4]
```diff
@@ -1,41 +1,44 @@
...
-3. Configure the `CONNECTED_VIA_BLUETOOTH` boolean correctly within definitions.py for your use case.
+1. Set the environment variables `FLASK_ENV=development` and `FLASK_APP=app.py`
+  1. If you use an IDE for running the code, you can usually set these in the configuration settings.
+  2. on Windows this is done by `set <varname>=<varvalue>`
+  3. On Mac this is done by `export <varname>=<valvalue>`
+2. Make sure the boolean variables in definitions.py are set correctly:
+  1. Set `DISTRIBUTION` to False if the frontend is run separately.
+  2. Set `CONNECTED_TO_PROTOTYPE` to True if you want the software to connect to the physical prototype.
+  3. Set `CONNECTION_VIA_BLUETOOH` to True if you want the software to connect to the physical prototype using bluetooth. (Note, you will have to have `CONNECTED_TO_PROTOTYPE` on True as well)
+  4. Set `RUNNING_ON_MAC` to True if you are running the software on a MAC or Linux machine
```
Comment (SanderSteeghs) [inline README.md:40]: You should not need to define this in `definitions.py` but 
`import os
os.system()` 
can already tell you if you are `Windows` `Darwin` or `Linux`

File: src/modules/PrototypeConnection.py [hunk 5]
```diff
@@ -195,58 +211,44 @@ def query(self, message: str) -> str:
...
+        nb = bluetooth.discover_devices(lookup_names=True)
+
+        # Loop over ports to check if device in question is in there.
+        for address, name in list(nb):
+            if bluetooth_device_name == name:
+                mac = address
+                Logger.log_info("MAC address to connect to: " + str(mac))
+                break
+
+        if mac is not None:
```
Comment (SanderSteeghs) [inline src/modules/PrototypeConnection.py:236]: I always prefer invert the `if` to reduce the complexity.
Basically I return early when I can and prevent the `else` keyword as much as possible

File: src/modules/PrototypeConnection.py [hunk 6]
```diff
@@ -59,29 +60,48 @@ def connect_with_config(self, fp_json: str):
...
-                            mac = self.known_bluetooth_mac_addresses[0]
-                        else:
-                            mac = self.find_in_reach_bluetooth_known_mac()
-                        port = self.get_spp_com_port(mac)
+                    if not RUNNING_ON_MAC:
+                        # If running on windows, then find the bluetooth port using pybluez
+                        port = self.find_bluetooth_port_windows(self.bluetooth_device_name)
                     else:
-                        port = self.com_port
+                        # Mac has standardized ports. The names of the ports include the name of the bluetooth chip,
```
Comment (SanderSteeghs) [inline src/modules/PrototypeConnection.py:69]: this function has a few too many indents and is hard to follow. Extract for example the part where you find the port when OS is mac.

File: README.md (hunk 4 above)
Comment (SanderSteeghs) [inline README.md:40]:
```suggestion
  4. Set `RUNNING_ON_MAC` to True if you are running the software on a MAC machine   
```
Linux is not supported right

File: README.md (hunk 2 above)
Comment (SanderSteeghs) [reply README.md:15]: `where` and `which` are commands in mac. 
How can the `$VIRTUAL_ENV` have a relative path? It is also not very reliable since one does not necessarily follow the `activate` the venv.
A more reliable method is to use hroncok's method within your `main.py`. This step can then also be removed from the readme

File: README.md [hunk 7]
```diff
@@ -1,41 +1,44 @@
...
 ### Development mode
 For running the application in development mode
-1. Set the environment variables `FLASK_ENV=development` and `FLASK_APP=app.py` (on Windows this is done by `set <varname>=<varvalue>`)
-2. Make sure the DISTRIBUTION boolean in definitions.py is set to `False`
-3. Configure the `CONNECTED_VIA_BLUETOOTH` boolean correctly within definitions.py for your use case.
+1. Set the environment variables `FLASK_ENV=development` and `FLASK_APP=app.py`
+  1. If you use an IDE for running the code, you can usually set these in the configuration settings.
+  2. on Windows this is done by `set <varname>=<varvalue>`
+  3. On Mac this is done by `export <varname>=<valvalue>`
+2. Make sure the boolean variables in definitions.py are set correctly:
```
Comment (SanderSteeghs) [inline README.md:36]: Idealy, we create some `yaml`/`json`/`xml` config file instead of `definitions.py`. Then you can remove the documentation from here and move it to that file instead. Advantage of that is that when you update/add/remove the parameters to you app, you are more inclined to change the documentation too. Otherwise the readme and code might go out-of-sync

File: README.md [hunk 8]
```diff
@@ -45,12 +48,6 @@ For running the application in development mode
 4. Change the links in the `index.html` referring to js/img/css files by adding the prefix `static/`
```
Comment (SanderSteeghs) [inline README.md:48]: Outside of the score of this PR, but we could consider automating this with a script.

File: src/modules/PrototypeConnection.py [hunk 9]
```diff
@@ -59,29 +60,48 @@ def connect_with_config(self, fp_json: str):
         # set found device if not in backend-debug mode
         if not self.debug:
```
Comment (SanderSteeghs) [inline src/modules/PrototypeConnection.py:61]: same as my comment below: if you invert this `if` statement, you can reduce the number of indents you need.
 ```suggestion
        if self.debug:
            Logger.log_info("PrototypeConnection initialized in debug mode.")
            return
        
        try:
            port = None 
        ...
```

**Review by SanderSteeghs — CHANGES_REQUESTED** [review]

File: README.md (hunk 2 above)
Comment (hendriklohse) [reply README.md:15]: hroncok's answer seems to be more robust indeed

File: src/routes/Routes.py [hunk 10]
```diff
@@ -15,302 +14,291 @@
...
-                            "coord": i,
-                            "amplitude": 255,
-                            "frequency": 300
-                        }
-                    ],
-                    "time": 1000
-                }
-            ]
+
+def init_views(app, dispatcher):
```
Comment (LourensT) [inline src/routes/Routes.py:18]: I might've missed why this refactor of putting everything in an init_views method is necessary, but nothing in the routes changed right? cus now it shows 500+ new lines

**Review by LourensT — CHANGES_REQUESTED** [review]
Code is fine imo, but I cannot test the refactor of prototypeconnection\s bluetooth part.  If that is tested the PR can be merged imo.

File: README.md (hunk 2 above)
Comment (rdvhoorn) [reply README.md]: I changed the line to just 'return the path of the virtualenvironment'. In addition i added some explanation at the top for more advanced users

File: src/routes/Routes.py (hunk 10 above)
Comment (rdvhoorn) [reply src/routes/Routes.py:18]: Its necessary for the way that flask loads servers. Its a bit weird, but this seemed to work best.

Comment (rdvhoorn) [reply src/routes/Routes.py:18]: But yes, that is the only thing that changed

**Review by SanderSteeghs — APPROVED** [review]
Nice job :-)

File: src/modules/PrototypeConnection.py [hunk 11]
```diff
@@ -64,7 +64,7 @@ def connect_with_config(self, fp_json: str):
         try:
             port = None
             if CONNECTED_VIA_BLUETOOTH:
-                if RUNNING_ON_MAC:
+                if os.name == 'nt': # Check if running on windows
```
[the remark this answers is not in the document]
Comment (rdvhoorn) [reply src/modules/PrototypeConnection.py]: you are correct lol good catch

[the remark this answers is not in the document]
Comment (hendriklohse) [reply src/modules/PrototypeConnection.py]: Changed it

**Review by hendriklohse — APPROVED** [review]

**Review by LourensT — APPROVED** [review]
LGTM
